lib/client-api/src/story_store.ts

```diff
@@ -383,8 +383,32 @@ export default class StoryStore {
     const loaders = [...this._globalMetadata.loaders, ...kindMetadata.loaders, ...storyLoaders];
 
     const finalStoryFn = (context: StoryContext) => {
-      const { passArgsFirst = true } = context.parameters;
-      return passArgsFirst ? (original as ArgsStoryFn)(context.args, context) : original(context);
+      const { args, argTypes, parameters } = context;
+      if (
+        argTypes &&
+        Object.values(argTypes).some(({ control }) => {
+          if (!control?.options) return false;
+          return Object.values(control.options).some((v: any) => v && typeof v === 'object');
+        })
+      ) {
+        once.warn(
+          'Objects cannot be used as values in control options. Use a `mapping` instead.\n\nMore info: https://storybook.js.org/docs/react/writing-stories/args#mapping-to-complex-arg-values'
+        );
+      }
+
+      const { passArgsFirst = true } = parameters;
+      if (args) {
+        const mapped = {
+          ...context,
+          args: Object.entries(args).reduce((acc, [key, val]) => {
+            const { mapping } = argTypes?.[key] || {};
+            acc[key] = mapping && val in mapping ? mapping[val] : val;
+            return acc;
+          }, {} as Args),
```

Comment:
```suggestion
          args: Object.fromEntries(
            Object.entries(args).map(([key, val]) => {
              const { mapping = {} } = argTypes?.[key] || {};
              return [key, val in mapping ? mapping[val] : val]
            })
          ),
```

🙏 ?

Reply:
The Storybook codebase doesn't support Object.fromEntries. And I don't really like it anyway because of the intermediate array. It's hella confusing.